metals/src/main/scala/scala/meta/internal/metals/callHierarchy/FindCallsResult.scala

```diff
@@ -0,0 +1,152 @@
+
+private[callHierarchy] sealed trait FindCallsResult[T, S <: FindCallsResult[
+  T,
+  _,
+]] {
+  def occurence: SymbolOccurrence
+  def definitionNameRange: lsp4j.Range
+  def fromRanges: List[lsp4j.Range]
+
+  protected def toLSP(
+      lspBuilder: (lsp4j.CallHierarchyItem, ju.List[lsp4j.Range]) => T
+  )(source: AbsolutePath, doc: TextDocument)(
+      builder: CallHierarchyItemBuilder,
+      visited: Array[String],
+      token: CancelToken,
+  )(implicit ec: ExecutionContext): Future[T] =
+    occurence match {
+      case SymbolOccurrence(_, symbol, _) =>
+        builder
+          .build(
+            source,
+            doc,
+            occurence,
+            definitionNameRange,
+            visited :+ symbol,
+            token,
+          )
+          .collect { case Some(item) =>
+            lspBuilder(
+              item,
+              fromRanges.asJava,
+            )
+          }
+    }
+
+  protected def aggregate(updatedFromRanges: List[lsp4j.Range]): S
+}
+
+/**
+ * Result of incoming calls finder.
+ *
+ * @param canBuildHierarchyItem The item that makes the call wrapped in CanBuildHierarchyItem.
+ * @param definitionNameRange The range that should be selected.
+ * @param fromRanges Ranges at which the calls appear.
+ */
+private[callHierarchy] case class FindIncomingCallsResult(
+    occurence: SymbolOccurrence,
+    definitionNameRange: lsp4j.Range,
+    fromRanges: List[lsp4j.Range],
+)(implicit ec: ExecutionContext)
+    extends FindCallsResult[
+      lsp4j.CallHierarchyIncomingCall,
+      FindIncomingCallsResult,
+    ] {
+
+  override protected def aggregate(
+      updatedFromRanges: List[lsp4j.Range]
+  ): FindIncomingCallsResult =
+    this.copy(fromRanges = updatedFromRanges)
+
+  private val anonToLSP: (AbsolutePath, TextDocument) => (
+      CallHierarchyItemBuilder,
+      Array[String],
+      CancelToken,
+  ) => Future[lsp4j.CallHierarchyIncomingCall] =
+    toLSP((from, fromRanges) =>
+      new lsp4j.CallHierarchyIncomingCall(from, fromRanges)
+    ) _
+
+  def toLSP(
+      source: AbsolutePath,
+      doc: TextDocument,
+      builder: CallHierarchyItemBuilder,
+      visited: Array[String],
+      token: CancelToken,
+  ): Future[lsp4j.CallHierarchyIncomingCall] =
+    anonToLSP(source, doc)(builder, visited, token)
+}
+
+/**
+ * Result of incoming calls finder.
+ *
+ * @param canBuildHierarchyItem The item that is called wrapped in CanBuildHierarchyItem.
+ * @param definitionNameRange The range that should be selected.
+ * @param fromRanges Ranges at which the item is called.
+ * @param source AbsolutePath where the item that is called is defined.
+ * @param doc TextDocument where the item that is called is defined.
+ */
+private[callHierarchy] case class FindOutgoingCallsResult(
+    occurence: SymbolOccurrence,
+    definitionNameRange: lsp4j.Range,
+    fromRanges: List[lsp4j.Range],
+    source: AbsolutePath,
+    doc: TextDocument,
+)(implicit ec: ExecutionContext)
```

Comment:
Execution context should be rather argument of `toLSP`. BTW, don't we have `toLsp` rather than `toLSP` convention?

Reply:
If you search for occurrences of `toLSP`, we have both. I can do an another PR to unify if necessary

Reply:
#4330